api/v1beta1/chaincode_types.go

```diff
@@ -0,0 +1,145 @@
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ChaincodePhase string
+
+const (
+	// ChaincodePhasePending create chaincode default state
+	ChaincodePhasePending ChaincodePhase = "ChaincodePending"
+	// ChaincodePhaseApproved the proposal is approved and the status of the chaincode
+	ChaincodePhaseApproved ChaincodePhase = "ChaincodeApproved"
+	// ChaincodePhaseUnapproved the proposal fails and the status of the chaincode
+	ChaincodePhaseUnapproved ChaincodePhase = "ChaincodeUnapproved"
+
+	// ChaincodePhaseRunning amazing
+	ChaincodePhaseRunning ChaincodePhase = "ChaincodeRunning"
+)
+
+type ChaincodeConditionType string
+
+const (
+	// ChaincodeCondPackaged chaincode packaged successfully
+	ChaincodeCondPackaged ChaincodeConditionType = "Packaged"
+	// ChaincodeCondInstalled chaincode code installed successfully
+	ChaincodeCondInstalled ChaincodeConditionType = "Installed"
+	// ChaincodeCondApproved chaincode definition approved
+	ChaincodeCondApproved ChaincodeConditionType = "Approved"
+	// ChaincodeCondCommitted chaincode commits to the chain
+	ChaincodeCondCommitted ChaincodeConditionType = "Committed"
+	// ChaincodeCondRunning chaincode is running, pod is in running state
+	ChaincodeCondRunning ChaincodeConditionType = "Running"
+	// ChaincodeCondError problems with these processes: package, install, approve, commit
+	ChaincodeCondError ChaincodeConditionType = "Error"
+	// ChaincodeCondDone process execution is complete and chaincode is running correctly
+	ChaincodeCondDone ChaincodeConditionType = "Done"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:deepcopy-gen=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=cc
+type Chaincode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
+	Spec ChaincodeSpec `json:"spec"`
+	// +operator-sdk:gen-csv:customresourcedefinitions.statusDescriptors=true
+	Status ChaincodeStatus `json:"status,omitempty"`
+}
+
+// ChaincodeList contains a list of Chaincode.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:deepcopy-gen=true
+type ChaincodeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Chaincode `json:"items"`
+}
+
+type ChaincodeSpec struct {
+	License License `json:"license"`
+
+	// Which channel does chaincode belong to.
+	Channel string `json:"channel"`
+	// chaincode id
+	// +kubebuilder:validation:Pattern:=`^[a-z][a-z0-9]{7,63}$`
+	ID string `json:"id"`
```

Comment:
https://hyperledger-fabric.readthedocs.io/en/release-2.5/chaincode_lifecycle.html#step-three-approve-a-chaincode-definition-for-your-organization
No rules found for `ID` definition. Is this regular definition reasonable?

Reply:
me neither... but chaincode approve just failed . maybe that's a bug of fabric chaincode. Let's mark this a bug/enhancement in the future.